### PR TITLE
8300176: URLEncoder/URLDecoder static fields should be private static final

### PR DESCRIPTION
**Pull request by DarraghClarke**
Made the `static` fields `private static final`, updated the naming as well to reflect this.

This meant I updated `URLDecoder` to set the `DEFAULT_ENCODING_NAME` itself since it had been previously getting it from `URLEncoder`.

Since these fields are only referenced from inside the classes nothing else needed updated

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300176](https://bugs.openjdk.org/browse/JDK-8300176): URLEncoder/URLDecoder static fields should be private static final

### Reviewers
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to [b08138d1](https://git.openjdk.org/jdk/pull/12122/files/b08138d1e31b360eb392a307b5299c4119d7f7c6)
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)
 * [Sergey Tsypanov](https://openjdk.org/census#stsypanov) (@stsypanov - Author)
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12122/head:pull/12122` \
`$ git checkout pull/12122`

Update a local copy of the PR: \
`$ git checkout pull/12122` \
`$ git pull https://git.openjdk.org/jdk pull/12122/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12122`

View PR using the GUI difftool: \
`$ git pr show -t 12122`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12122.diff">https://git.openjdk.org/jdk/pull/12122.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/net/URLDecoder.java [hunk 1]
```diff
@@ -89,7 +91,11 @@ public class URLDecoder {
     private URLDecoder() {}
 
     // The default charset
-    static String dfltEncName = URLEncoder.dfltEncName;
+    private static final String DEFAULT_ENCODING_NAME;
+
+    static {
+        DEFAULT_ENCODING_NAME = StaticProperty.fileEncoding();
```
Comment (stsypanov) [inline src/java.base/share/classes/java/net/URLDecoder.java:97]: Wouldn't it be more brief to move initializer to field declaration?

**Review by cl4es — APPROVED** [review]
This looks OK to me, though as @stsypanov notes initialization of `DEFAULT_ENCODING_NAME` can be further simplified.

There's an existing microbenchmark, `TestEncodeDecode`, that might be affected by these changes. It'd be interesting to get some data on before/after. 

Pre-existing, but I'll also note that the `URLEncoder::DONT_NEED_ENCODING` could be constructed as `new BitSet(128)`: all set bits are in the ASCII range (0-127). Could be a win to trim it down.

File: src/java.base/share/classes/java/net/URLEncoder.java [hunk 2]
```diff
@@ -119,25 +119,25 @@ public class URLEncoder {
          *
          */
 
-        dontNeedEncoding = new BitSet(256);
+        DONT_NEED_ENCODING = new BitSet(256);
         int i;
         for (i = 'a'; i <= 'z'; i++) {
```
Comment (cl4es) [inline src/java.base/share/classes/java/net/URLEncoder.java:124]: These three for-loops could be simplified to: 
```java
        DONT_NEED_ENCODING.set('a', 'z' + 1);
        DONT_NEED_ENCODING.set('A', 'Z' + 1);
        DONT_NEED_ENCODING.set('0', '9' + 1);
```

File: src/java.base/share/classes/java/net/URLDecoder.java [hunk 3]
```diff
@@ -108,7 +114,7 @@ public static String decode(String s) {
         String str = null;
 
         try {
-            str = decode(s, dfltEncName);
+            str = decode(s, DEFAULT_ENCODING_NAME);
```
Comment (schlosna) [inline src/java.base/share/classes/java/net/URLDecoder.java:117]:
```suggestion
            str = decode(s, StaticProperty.fileEncoding());
```

File: src/java.base/share/classes/java/net/URLDecoder.java [hunk 4]
```diff
@@ -89,7 +91,11 @@ public class URLDecoder {
     private URLDecoder() {}
 
     // The default charset
-    static String dfltEncName = URLEncoder.dfltEncName;
+    private static final String DEFAULT_ENCODING_NAME;
+
+    static {
+        DEFAULT_ENCODING_NAME = StaticProperty.fileEncoding();
+    }
```
Comment (schlosna) [inline src/java.base/share/classes/java/net/URLDecoder.java:99]: Can we just in-line the use of `StaticProperty.fileEncoding()` to the call site as `StaicProperty.FILE_ENCODING` is already stored in a static final field

```suggestion
```

https://github.com/openjdk/jdk/blob/77a50105f0c4a4cb7286dda13c633f1e69295210/src/java.base/share/classes/jdk/internal/util/StaticProperty.java#L54-L74

File: src/java.base/share/classes/java/net/URLEncoder.java [hunk 5]
```diff
@@ -77,9 +77,9 @@
  * @since   1.0
  */
 public class URLEncoder {
-    static BitSet dontNeedEncoding;
-    static final int caseDiff = ('a' - 'A');
-    static String dfltEncName;
+    private static final BitSet DONT_NEED_ENCODING;
+    private static final int CASE_DIFF = ('a' - 'A');
+    private static final String DEFAULT_ENCODING_NAME;
```
Comment (schlosna) [inline src/java.base/share/classes/java/net/URLEncoder.java:82]: Similar to above, can we inline usages?

```suggestion
```

File: src/java.base/share/classes/java/net/URLEncoder.java [hunk 6]
```diff
@@ -119,25 +119,25 @@ public class URLEncoder {
...
-        dontNeedEncoding.set('_');
-        dontNeedEncoding.set('.');
-        dontNeedEncoding.set('*');
+        DONT_NEED_ENCODING.set('-');
+        DONT_NEED_ENCODING.set('_');
+        DONT_NEED_ENCODING.set('.');
+        DONT_NEED_ENCODING.set('*');
 
-        dfltEncName = StaticProperty.fileEncoding();
+        DEFAULT_ENCODING_NAME = StaticProperty.fileEncoding();
```
Comment (schlosna) [inline src/java.base/share/classes/java/net/URLEncoder.java:140]:
```suggestion
```

File: src/java.base/share/classes/java/net/URLEncoder.java [hunk 7]
```diff
@@ -162,7 +162,7 @@ public static String encode(String s) {
         String str = null;
 
         try {
-            str = encode(s, dfltEncName);
+            str = encode(s, DEFAULT_ENCODING_NAME);
```
Comment (schlosna) [inline src/java.base/share/classes/java/net/URLEncoder.java:165]:
```suggestion
            str = encode(s, StaticProperty.fileEncoding());
```

**Review by schlosna — COMMENTED** [review]
Curious if we could remove a couple of the static fields

**Comment by DarraghClarke** [comment]
I pushed changes there based on the comments, and to touch on some points brought up

Personally I'd prefer to keep `DEFAULT_ENCODING_NAME` just for readability, though I'm not incredibly pushed on that if folk think it'd be better to just replace its usages with a direct call.

And as for testing, I ran some benchmarks on my local machine that did show a speed up after the changes, but this was my first time using JMH so I was wondering if you have any pointers @cl4es as to what settings/how many iterations should I run before I can be confident in any results it gives?

**Review by Michael-Mc-Mahon — APPROVED** [review]
Looks fine to me

**Review by stsypanov — APPROVED** [review]

**Review by dfuch — APPROVED** [review]
LGTM

**Comment by cl4es** [comment]
> And as for testing, I ran some benchmarks on my local machine that did show a speed up after the changes, but this was my first time using JMH so I was wondering if you have any pointers @cl4es as to what settings/how many iterations should I run before I can be confident in any results it gives?

As a rule of thumb if you get the approximate same results when running the same benchmark back to back and the difference between your baseline and test results is decidedly larger than the sum of the errors then you can have a reasonably high confidence in the results.

If errors are too high for comfort (or effect too small) you should first consider if your benchmarking system can be made quieter (e.g. turning off browsers, slack, IDEs, turn off hyperthreading, turbo boost, apply firmware updates, put the box in a freezer, don't move, don't make a sound...!), then if you're still not getting comfortably consistent results you can increase the amount of forks, warmup and measurement iterations you're running with.

**Review by jaikiran — APPROVED** [review]

**Pull request closed without merging**